controller/kubernetes_controller.go

```diff
@@ -0,0 +1,409 @@
+			kc.enqueuePVDeletion(pv)
+		},
+	})
+
+	podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			pod := obj.(*v1.Pod)
+			kc.enqueuePodChange(pod)
+		},
+		UpdateFunc: func(old, cur interface{}) {
+			curPod := cur.(*v1.Pod)
+			kc.enqueuePodChange(curPod)
+		},
+		DeleteFunc: func(obj interface{}) {
+			pod := obj.(*v1.Pod)
+			kc.enqueuePodChange(pod)
+		},
+	})
+
+	return kc
+}
+
+func (kc *KubernetesController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer kc.queue.ShutDown()
+
+	logrus.Infof("Start kubernetes controller")
+	defer logrus.Infof("Shutting down kubernetes controller")
+
+	if !controller.WaitForCacheSync("kubernetes", stopCh, kc.pvStoreSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(kc.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (kc *KubernetesController) worker() {
+	for kc.processNextWorkItem() {
+	}
+}
+
+func (kc *KubernetesController) processNextWorkItem() bool {
+	key, quit := kc.queue.Get()
+
+	if quit {
+		return false
+	}
+	defer kc.queue.Done(key)
+
+	err := kc.syncKubernetesStatus(key.(string))
+	kc.handleErr(err, key)
+
+	return true
+}
+
+func (kc *KubernetesController) handleErr(err error, key interface{}) {
+	if err == nil {
+		kc.queue.Forget(key)
+		return
+	}
+
+	if kc.queue.NumRequeues(key) < maxRetries {
+		logrus.Warnf("Error syncing Longhorn volume status %v: %v", key, err)
+		kc.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logrus.Warnf("Dropping Longhorn volume %v out of the queue: %v", key, err)
+	kc.queue.Forget(key)
+}
+
+func (kc *KubernetesController) syncKubernetesStatus(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "fail to sync %v", key)
+	}()
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	err = kc.cleanupForPVDeletion(name)
+	if err != nil {
+		return err
+	}
+
+	pv, err := kc.kubeClient.CoreV1().PersistentVolumes().Get(name, metav1.GetOptions{})
+	if err != nil && !IsErrorNotFound(err) {
+		return errors.Wrapf(err, "Error getting Persistent Volume %s", name)
+	}
+	volumeName, err := kc.validatePersistentVolume(pv)
+	if err != nil {
+		return err
+	}
+	if volumeName == "" {
+		return nil
+	}
+
+	volume, err := kc.ds.GetVolume(volumeName)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			logrus.Infof("Longhorn volume %v has been deleted", volumeName)
+		}
+		return nil
+	}
+
+	// existing volume may be reused by other pv
+	ks := &volume.Status.KubernetesStatus
+	if ks.PVName != name {
+		ks.Namespace = ""
+		ks.PVCName = ""
+		ks.PodName = ""
+		ks.PodStatus = ""
+		ks.WorkloadName = ""
+		ks.WorkloadType = ""
+		ks.LastPVCRefAt = ""
+		ks.LastPodRefAt = ""
+	}
+
+	// PVC gets bound with PV when PVStatus becomes Bound
+	if pv.Spec.ClaimRef != nil && pv.Status.Phase == v1.VolumeBound && ks.PVStatus != string(pv.Status.Phase) {
```

Comment:
The flow starting here is really jumpy.

The condition branches are a great way to help you think all the different cases. But here it's hard to tell if the logic here covers all the cases. Tons of things can fall into cracks here.

Is that possible `pv.Spec.ClaimRef == nil && pv.Status.Phase == v1.VolumeBound`? And what if `PVStatus == Phase` (which is `Bound`) but ClaimRef changed? Maybe it's covered, may not, but it's not easy to tell from reading this part of code.

This part should be rewritten to clarlify the logic flow.